File: sensors.py
import time, os
import datetime
import rrdtool
import socket 
import json
import couchdb
import smbus
couch = couchdb.Server('http://192.168.8.1:5984/')
db = couch['sensors']
beauty_ip = "10.70.60.1"
beauty_port = 2003
path = "/sys/bus/w1/devices/"
tempSensor = { 'beagle' : '10-0008029674ee', 'out' : '10-000802964c0d', 'cylinder' : '10-000802961f0d', 'stove' : '10-00080296978d', 'canal' : '28-000004ee99a8' }
acs714=0.066
acs709=0.028

i2cBus = smbus.SMBus(1)
#
Vcc=5.0
mid=Vcc/2
step=Vcc/1024

def timestamp():
    timestamp = json.dumps(datetime.datetime.now().isoformat()).split('"')[1]
    return timestamp

def getTemp(sensor,address):
    if os.path.isfile(path+address+"/w1_slave"):
        try:
            raw = open(path+address+"/w1_slave", "r").read()
            if 'YES' in raw:
                current_temp = float(raw.split("t=")[-1])/1000
                data = { sensor : current_temp }
                return data
            else:
                return 'crc_error'
        except IOError:
            print "%s %s is offline" % (sensor,address)

def get1wireValues(sensors):
    data = {}
    for k,v in sensors.iteritems():
        if os.path.isfile("/sys/bus/w1/devices/"+v+"/w1_slave"):
            try:
                raw = open(path+v+"/w1_slave", "r").read()
                if 'YES' in raw:
                    current_temp = float(raw.split("t=")[-1])/1000
                    data[k] = current_temp 
            except IOError:
                print "%s %s is offline" % (sensor,address)
    return data
def getADCvalues(address):
    sensor=0
    adc_raw={'adc0': [],'adc1': [],'adc2':[],'adc3': [],'adc4': [],'adc5': [],'adc6':[],'adc7': [],'adc8': [],'adc9': [] }
    for reading in range(10):    
        i2cBus.write_byte(0x28, 0x00)
        adc = i2cBus.read_i2c_block_data(0x28, 0x00, 0x14)
        pairs = zip(adc[::2], adc[1::2])
        sensor=0
        for pair in pairs:
            a,b = pair
            value = a*256+b
            adc_raw['adc'+str(sensor)].append(value)
            sensor+=1
    #print adc_raw
    for i in range(10):
        adc=adc_raw['adc'+str(i)]
        adc.remove(max(adc))
        adc.remove(min(adc))
        adc_raw['adc'+str(i)]=sum(adc)/len(adc) 
    return adc_raw

def voltage(raw,name):
    """ get a raw value and convert it to 12v"""
    voltage= raw * 0.015357
    return { 'name': name,'raw': raw ,'volts': voltage }

def current(raw,name,step):
    """convert an int (0-1024) to Amps """
    values= []
    volts=(raw)*(Vcc/1016)
    amps = (volts-mid)/step
    return { 'name': name,'raw': raw ,'volts': volts ,'amps': amps}
    
def toGraphite(category,name,value,timestamp):
    graph  = ".".join(['boat',category,name])
    packet = " ".join([graph, str(value), timestamp])
    sock.sendto(packet, (beauty_ip,beauty_port))
    print packet

def toCouchdb(cat,data):
    doc = { "_id": now , "type" : cat, "data": data }
    print doc
    db.save(doc)


sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) 
now = str(time.time())

i2c1 = getADCvalues('0x28')
print i2c1

solar = current(i2c1['adc0'],'solar',acs714)
genny = current(i2c1['adc1'],'genny',acs709)
bat1  = voltage(i2c1['adc3'],'batteries')
cat='current'
now = str(time.time())

toGraphite('current',solar['name'],solar['amps'],now)
toCouchdb(cat,solar)
now = str(time.time())
toGraphite('current',genny['name'],genny['amps'],now)
toCouchdb(cat,genny)

now = str(time.time())
toGraphite('volts','batteries',bat1['volts'],now)
toCouchdb('volts',bat1)

now = str(time.time())
temps = get1wireValues(tempSensor)
toCouchdb('temp',temps)
#toGraphite('temp',bat1['name'],bat1['volts'],now)
#print temps



# while True:
#     now = str(time.time())
#     for k,v in tempSensor.iteritems():
#         temp_data=getTemp(k,v)
#         if 'crc_error' not in data[k]:
#             toGraphite('temp',k,temp,now)

#             print "%s  : %s C" %(k, str(temp))
#         toCouchdb(now,'temp','1wire',data)
#         print ""
#     time.sleep(60)
#     print "----------"

    

File: sensors.rb
#!/usr/bin/ruby
require 'json'
require 'time'
# need 'i2c' gem installed
require "i2c/i2c"
require "i2c/backends/i2c-dev"
require 'couchrest'
require 'simple-graphite'
require 'socket'
require 'uri'

module Boattr
  class Sensors
    @@data = []
    @@device 
    @@OWdevices = []
    attr_reader :location, :i2cAddress, :i2cBus, :data, :basename
    def initialize(params)
      @@basename    = params['basename']
      @i2cAddress   = params['i2cAddress']
      @@device      = ::I2C.create(params['i2cBus'])
      self.read()
      self.onewire()
    end
    def readold
      @data = []
      # read 20 bytes from slave, convert to decimals, and finaly in 10bit values.
      @adc = @@device.read(i2cAddress, 0x14, 0x00).unpack('C*').map {|e| e.to_s 10}
      #slice the 20 byte array into pairs (MSB,LSB) and convert. 
      @adc.each_slice(2) {|a| @@data << a[0].to_i*256+a[1].to_i}
      return @@data
    end
    def read
      @i2cIter      = 12
      @data = Array.new(10) {Array.new }
      @samples = []
      @i2cIter.times do #we take @i2cIter samples
        # read 20 bytes from slave, convert to decimals, and finaly in 10bit values.
        @adc = @@device.read(i2cAddress, 0x14, 0x00).unpack('C*').map {|e| e.to_s 10}
        sleep(0.1)
        #slice the 20 byte array into pairs (MSB,LSB) and convert decimal. 
        @adc.each_slice(2) {|a| @samples << a[0].to_i*256+a[1].to_i}
        #p @samples
        @data.each_with_index() {|d,i| d << @samples[i]}
        @samples = []
      end
      #take the average (array.inject(:+) ? )
      @data.each_with_index() do |d,i|
        # sort and remove max and min
        d = d.sort
        d.pop 
        d.shift
        @@data << d.inject{|sum,x| sum + x }/(@i2cIter-2)
      end
      #p @@data
      return  @@data
    end

    def voltage(name,address)
      @name    = name
      @raw     = @@data[address]
      @volts = @raw * 0.015357
      return { 'name' => @name, 'type' => 'volts', 'raw' => @raw, 'value' => @volts.round(3) }
    end
    def current(name,address)
      @name    = name
      @raw     = @@data[address]
      @volts   = (@raw*0.004887)
      @amps    = (@volts-2.5)/0.066 
      return { 'name' => @name, 'type' => 'current', 'raw' => @raw, 'value' => @amps.round(3)} 
    end
    def onewire()
      @basedir  = '/sys/bus/w1/devices/'
      Dir.chdir(@basedir)
      @@OWdevices = Dir['*-*']
      return @@OWdevices
    end
    def temperature(name,address)
      @name     = name
      @address  = address
      @basedir  = '/sys/bus/w1/devices/'
      if @@OWdevices.include?(@address) then
        file = File.open("#{@basedir}/#{address}/w1_slave",'r')
        if file.readline().include?('YES') then ## Is CRC valid in the first line? lets read the second and extract the temp
          @temp = file.readline().split()[-1].split('=')[-1].to_i/1000.0
          return { 'name' => @name, 'address' => @address, 'type' => 'temp', 'value' => @temp.round(3) }
        end
      end
    end
    def waterlevel(name,address)
      @name     = name
      @address  = address
      @raw      = @@data[address]
      @volts    = @raw*0.00472
      p @volts
      return { 'name' => @name, 'type' => 'water', 'value' => @raw }
    end
  end
  class Data  
    def initialize(params)
      @graphite = params['graphite']
      @couchdb  = params['couchdb']
      @@basename = params['basename']
      @g     = Graphite.new({:host => "#{@graphite}", :port => 2003})
      @sensorsdb = CouchRest.database!("http://#{@couchdb}:5984/#{@@basename}-sensors")
      @statsdb   = CouchRest.database!("http://#{@couchdb}:5984/#{@@basename}-stats")
    end
    def to_db(sensor_data)
      @data  = sensor_data
      @data.each() do |x|
        if x.nil? then 
          next
        end
        p x
        @doc ={ "_id" => now() }.merge x
        @sensorsdb.save_doc(@doc)
      end
    end
    def create_views(sensor_data)
      @sensorsdb.save_doc(
                          {
                            "_id" => "_design/current", 
                            :views => {
                              :solar => {
                                :map => "function(doc) {  if (doc.name == \"solar\") {  emit(doc._id, doc.value);  }}"
                              },
                              :ring => {
                                :map => "function(doc) {  if (doc.name == \"ring\") {  emit(doc._id, doc.value);  }}"
                              },
                              :fridge => {
                                :map => "function(doc) {  if (doc.name == \"fridge\") {  emit(doc._id, doc.value);  }}"
                              },
                              :lights => {
                                :map => "function(doc) {  if (doc.name == \"lights\") {  emit(doc._id, doc.value);  }}"
                              }

                            }

                          })
    end
    def amphours(name,hours)
      @sum   = 0
      @name  = name
      @hours = hours
      @from = Time.now().to_i-hours*60*60
      @view = URI.escape("current/#{@name}?startkey=\"#{@from}\"")
      @data = @sensorsdb.view(@view)
      @rows_returned = @data['total_rows'] - @data['offset'] 
      @rows       = @data['rows']
      #p  @rows_returned
      @rows.each() do |r|
        @sum+=r['value']
      end
      @amph = @sum/(@rows_returned/@hours)
      return @amph
    end
    def to_graphite(sensor_data)
      @base  = @@basename
      @data  = sensor_data
      #p "#{@base}.#{@type}.#{@name} #{@value} #{@g.time_now}"
      @data.each() do |x|
        if x.nil? then 
          next
        end
        @type  = x['type']
        @name  = x['name']
        @value = x['value']
        @g.push_to_graphite do |graphite|
          graphite.puts "#{@base}.#{@type}.#{@name} #{@value} #{@g.time_now}"
        end
      end
    end
    def now
      #used by to_db()
      return Time.now.to_f.round(2).to_s
    end
    def to_dashboard(sensor_data)
      @name =  @@basename
      HTTParty.post('http://localhost:3030/widgets/karma',
                    :body => { auth_token: "YOUR_AUTH_TOKEN", current: 1000 }.to_json)
    end
  end

end
